### PR TITLE
SMILESWidget: Store SMILES in StructureData extras

### LINKED ISSUE #264: Add smiles code to StructureData.extras in SmilesWidget
Just a small feature request, I think it would be nice for structures generated from SMILES code to have that info preserved on the StructureData node. I am an aiida noob, but node extras seems to be a good place to put it? I am happy to submit a PR if there is interest in this functionality.

PS: Kudos to the aiida team, have been playing with AiidaLab for a while at it was surprisingly easy to get going thanks to the good documentation and clear code. Thanks! 
(I am working on an AiidaLab app for generating UV/Vis spectra).

### PR DESCRIPTION
**Pull request by danielhollas**
Hi folks, here's a PR as discussed in #264. Happy to make any changes. LMK if there is a good way to write a test for this.

Closes #264

### CODE REVIEW

File: aiidalab_widgets_base/structures.py [hunk 1]
```diff
@@ -653,25 +659,28 @@ def __init__(self, title=""):
...
         atoms.cell = np.ptp(atoms.positions, axis=0) + 10
         atoms.center()
+        # We're attaching this info so that it
+        # can be later stored as an extra on AiiDA Structure node.
+        atoms.info["smiles"] = smiles
 
         return atoms
 
-    def _pybel_opt(self, smile, steps):
+    def _pybel_opt(self, smiles, steps):
```
Comment (danielhollas) [inline aiidalab_widgets_base/structures.py:675]: Unrelated: `s/smile/smiles` to make the code consistent.

File: aiidalab_widgets_base/structures.py [hunk 2]
```diff
@@ -265,6 +265,12 @@ def _convert_to_structure_node(self, structure):
 
         # If the input_structure trait is set to Atoms object, structure node must be created from it.
         if isinstance(structure, Atoms):
+            # If the Atoms object was created by SmilesWidget,
+            # attach its SMILES code as an extra.
+            if "smiles" in structure.info:
+                structure_node = structure_node_type(ase=structure)
+                structure_node.set_extra("smiles", structure.info["smiles"])
+                return structure_node
             return structure_node_type(ase=structure)
```
Comment (yakutovicha) [inline aiidalab_widgets_base/structures.py:274]: Just a bit shorter.

```suggestion
            structure_node = structure_node_type(ase=structure)
            if "smiles" in structure.info:
                structure_node.set_extra("smiles", structure.info["smiles"])
            return structure_node
```

File: aiidalab_widgets_base/structures.py [hunk 3]
```diff
@@ -653,25 +659,28 @@ def __init__(self, title=""):
 
         super().__init__([self.smiles, self.create_structure_btn, self.output])
 
-    def make_ase(self, species, positions):
+    def make_ase(self, species, positions, smiles):
```
Comment (yakutovicha) [inline aiidalab_widgets_base/structures.py:662]: We are changing the API of several methods, which is not so good, in principle. But, they should not be called outside of this class, so I would suggest making them "internal":

```suggestion
    def _make_ase(self, species, positions, smiles):
```

File: aiidalab_widgets_base/structures.py [hunk 4]
```diff
@@ -685,30 +694,30 @@ def _pybel_opt(self, smile, steps):
         f_f.GetCoordinates(pbmol.OBMol)
         species = [chemical_symbols[atm.atomicnum] for atm in pbmol.atoms]
         positions = np.asarray([atm.coords for atm in pbmol.atoms])
-        return self.make_ase(species, positions)
+        return self.make_ase(species, positions, smiles)
```
Comment (yakutovicha) [inline aiidalab_widgets_base/structures.py:697]:
```suggestion
        return self._make_ase(species, positions, smiles)
```

File: aiidalab_widgets_base/structures.py [hunk 5]
```diff
@@ -685,30 +694,30 @@ def _pybel_opt(self, smile, steps):
...
+        mol = Chem.MolFromSmiles(smiles)
         mol = Chem.AddHs(mol)
 
         AllChem.EmbedMolecule(mol, maxAttempts=20, randomSeed=42)
         AllChem.UFFOptimizeMolecule(mol, maxIters=steps)
         positions = mol.GetConformer().GetPositions()
         natoms = mol.GetNumAtoms()
         species = [mol.GetAtomWithIdx(j).GetSymbol() for j in range(natoms)]
-        return self.make_ase(species, positions)
+        return self.make_ase(species, positions, smiles)
```
Comment (yakutovicha) [inline aiidalab_widgets_base/structures.py:713]:
```suggestion
        return self._make_ase(species, positions, smiles)
```

File: aiidalab_widgets_base/structures.py [hunk 6]
```diff
@@ -685,30 +694,30 @@ def _pybel_opt(self, smile, steps):
...
 
-    def mol_from_smiles(self, smile, steps=10000):
+    def mol_from_smiles(self, smiles, steps=10000):
```
Comment (yakutovicha) [inline aiidalab_widgets_base/structures.py:715]: Please change the method name in the other places too.
```suggestion
    def _mol_from_smiles(self, smiles, steps=10000):
```

**Review by yakutovicha — CHANGES_REQUESTED** [review]
Thanks, @danielhollas, overall looks good to me. I just have a couple of minor suggestions.

**Comment by danielhollas** [comment]
@yakutovicha thanks for the review! I implemented all the proposed changes.

**Review by yakutovicha — APPROVED** [review]

**Comment by yakutovicha** [comment]
Thanks a lot, @danielhollas!

**Pull request merged**